tools/src/main/java/org/apache/kafka/trogdor/workload/SustainedConnectionWorker.java

```diff
@@ -0,0 +1,563 @@
+        }
+
+        @Override
+        public void refresh() {
+            try {
+                if (this.client == null) {
+                    // Housekeeping to track the number of opened connections.
+                    SustainedConnectionWorker.this.totalMetadataConnections.incrementAndGet();
+
+                    // Create the admin client connection.
+                    this.client = AdminClient.create(this.props);
+                }
+
+                // Fetch some metadata to keep the connection alive.
+                this.client.describeCluster().nodes().get();
+
+            } catch (Throwable e) {
+                // Set the admin client to be recreated on the next cycle.
+                this.closeQuietly();
+
+                // Housekeeping to track the number of opened connections and failed connection attempts.
+                SustainedConnectionWorker.this.totalMetadataConnections.decrementAndGet();
+                SustainedConnectionWorker.this.totalMetadataFailedConnections.incrementAndGet();
+                SustainedConnectionWorker.log.error("Error while refreshing sustained AdminClient connection", e.getMessage());
+            }
+
+            // Schedule this again and set to not in use.
+            this.nextUpdate = System.currentTimeMillis() + this.refreshRate;
+            this.inUse = false;
+        }
+
+        @Override
+        protected void closeQuietly() {
+            Utils.closeQuietly(this.client, "AdminClient");
+            this.client = null;
+        }
+    }
+
+    private class ProducerSustainedConnection extends ClaimableConnection {
+
+        // These variables are used to maintain the connection itself.
+        private KafkaProducer<byte[], byte[]> producer;
+        private List<TopicPartition> partitions;
+        private String topicName;
+        private Iterator<TopicPartition> partitionsIterator;
+        private PayloadIterator keys;
+        private PayloadIterator values;
+
+        // This variable is used to maintain the scheduling of the connection keep-alive.
+        private long refreshRate;
+
+        // This variable is used to maintain the connection properties.
+        private final Properties props;
+
+        ProducerSustainedConnection() {
+
+            // These variables are used to maintain the connection itself.
+            this.producer = null;
+            this.partitions = null;
+            this.topicName = SustainedConnectionWorker.this.spec.topicName();
+            this.partitionsIterator = null;
+            this.keys = new PayloadIterator(SustainedConnectionWorker.this.spec.keyGenerator());
+            this.values = new PayloadIterator(SustainedConnectionWorker.this.spec.valueGenerator());
+
+            // This variable is used to maintain the scheduling of the connection keep-alive.
+            this.refreshRate = SustainedConnectionWorker.this.spec.refreshRateMs();
```

Comment:
I suggest passing all of the variables we access in this constructor via`SustainedConnectionWorker.this.spec` to be switched to parameters we pass upon instantiation

Reply:
I'd rather not.  That's why I was originally passing in the spec to begin with.

Reply:
Why did you decide against passing in the spec or the arguments in the end?
It makes the class more unit-testable (not that we have unit tests) and it's generally easier to read I find

Reply:
https://github.com/apache/kafka/pull/7289#discussion_r320514522

Reply:
Sorry about that. In the end I think I prefer passing it in but I don't have a strong opinion

Reply:
I will leave it as-is for now, if we want to change it, it's a simple fix.